src/ecmult_const_impl.h

```diff
@@ -228,4 +228,58 @@ static void secp256k1_ecmult_const(secp256k1_gej *r, const secp256k1_ge *a, cons
     secp256k1_fe_mul(&r->z, &r->z, &Z);
 }
 
+static int secp256k1_ecmult_const_xonly(secp256k1_fe* r, const secp256k1_fe *n, const secp256k1_fe *d, const secp256k1_scalar *q, int bits, int known_on_curve) {
+
+    /* This algorithm is a generalization of Peter Dettman's technique for
+     * avoiding the square root in a random-basepoint x-only multiplication
+     * on a Weierstrass curve:
+     * https://mailarchive.ietf.org/arch/msg/cfrg/7DyYY6gg32wDgHAhgSb6XxMDlJA/
+     */
+    secp256k1_fe g, i;
+    secp256k1_ge p;
+    secp256k1_gej rj;
+
+    /* Compute g = (n^3 + B*d^3). */
+    secp256k1_fe_sqr(&g, n);
+    secp256k1_fe_mul(&g, &g, n);
+    if (d) {
+        secp256k1_fe b;
+        secp256k1_fe_sqr(&b, d);
+        secp256k1_fe_mul(&b, &b, d);
+        secp256k1_fe_mul(&b, &b, &secp256k1_fe_const_b);
+        secp256k1_fe_add(&g, &b);
+        if (!known_on_curve) {
+            secp256k1_fe c;
+            secp256k1_fe_mul(&c, &g, d);
+            if (secp256k1_fe_jacobi_var(&c) < 0) return 0;
+        }
+    } else {
+        secp256k1_fe_add(&g, &secp256k1_fe_const_b);
```

Comment:
not this PR: Now that I see this, I wonder if we should get rid of `secp256k1_fe_const_b` and simply call `secp256k1_fe_set_int`... I don't think it makes a measurable difference in performance (or it may even be faster).